### ISSUE TITLE
Predict on batch being able to do the prediction with inputs of different first shapes

### ISSUE THREAD

**Issue by tanguycdls** [tensorflow/tensorflow#42280]
**System information**
- TensorFlow version (you are using): 2.3
- Are you willing to contribute it (Yes/No): yes

**Describe the feature and the current behavior/state.**
For a recommender system usage we want to be able to have a keras model that takes inputs with different first dimensions, for example one input with N rows and another with one row.
That feature was also asked here https://github.com/keras-team/keras/issues/13646 for another use case.

Until https://github.com/tensorflow/tensorflow/commit/56a0ce87911236765633d2a873e706ebc6401ef9 predict on batch did not check the cardinality of tensors before creating the dataset so our usage fits in the Keras api.
but now there is an additional _check_data_cardinality(data):
https://github.com/tensorflow/tensorflow/commit/56a0ce87911236765633d2a873e706ebc6401ef9#diff-f8dd40712ac721c1b363e1a1ec44c1a3R1523
that will break if you pass two inputs one with one row and one with N rows. 

**Will this change the current api? How?**
Provide a way to disable that check, either making it a warning or removing it. 

**Who will benefit with this feature?**
Users with non static input shapes, recommenders systems where we want to take into advantage the fact that we do a prediction on 1 movie for all the users to use broadcasting instead of running N times the same embedding. 
**Any Other info.**

**Comment by bela127** [comment]
Hello,

you could just use the **call** method instead of the **predict** or **predict_batch** method.
This would make your code even shorter, because you can write:
`output = model(input)` instead of `output = model.predict(input)`

maybe it will solve your issue?

**Comment by gowthamkpr** [comment]
@tanguycdls As mentione in the above comment, did you try using the call method?

**Comment by tanguycdls** [comment]
hi thanks @bela127 @gowthamkpr yes it works but for performance reason i want to graph it. I could wrap it into a tf function but my input is a dict of ragged tensor and those are currently not supported well while in a dict.
In the mean time i found another trick: if you use a generator and predict instead the check does not happen as well.

I just find that check odd at such high level of the api: for datasets the check makes sense since we might want to batch it so we would need to split the data in the first dim: But for predict_on_batch there is only one batch and if there is a shape issue I think Tensorflow in the graph will break with a clear error.

**Comment by blackyang** [comment]
+1 to what @tanguycdls mentioned, in predict_on_batch we don't need such check, can we just raise an error if there is any issue following the "let it crash" philosophy?

**Comment by sushreebarsa** [comment]
@tanguycdls  It looks like you are using an older Version of Tensorflow (2.3). Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version (2.4.1 or 2.5.0) and let us know if the issue still persists? Thanks!

**Comment by tanguycdls** [comment]
@sushreebarsa Yes the issue persists, the workaround is to create a custom tf dataset and then we no longer call the check_data_cardinality function. 
This is why I believe the behavior is strange, if you create a dataset using predict directly it breaks, if you create beforehand the tf dataset it works.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
